Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs

```diff
@@ -221,49 +299,210 @@ private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs fi
-                dimensions.Add("operation", this.GetDimCappedString(operation, this.operationValues));
+                refinedProblemId = "MaxProblemIdValues";
             }
 
-            var metric = this.metricManager.CreateMetric("Exceptions Thrown", dimensions);
+            dimensions.Add("problemId", refinedProblemId);
 
-            if (wasTracked)
+            if (SdkInternalOperationsMonitor.IsEntered())
             {
-                metric.Track(0);
+                dimensions.Add("operationName", "AI (Internal)");
             }
             else
             {
-                metric.Track(1);
+                if (getOperationName == true)
+                {
+                    exceptionTelemetry = new ExceptionTelemetry(exception);
+                    this.telemetryClient.Initialize(exceptionTelemetry);
+                    operationName = exceptionTelemetry.Context.Operation.Name;
+                }
+
+                if (string.IsNullOrEmpty(operationName) == false)
+                {
+                    refinedOperationName = this.GetDimCappedString(operationName, this.operationNameValues, OperationNameCacheSize);
+
+                    dimensions.Add("operationName", refinedOperationName);
+                }
             }
+
+            this.SendException(refinedOperationName, refinedProblemId, exceptionTelemetry, exception);
+
+            var metric = this.metricManager.CreateMetric("Exceptions thrown", dimensions);
+
+            metric.Track(1);
         }
 
-        private string GetDimCappedString(string dimensionValue, ConcurrentDictionary<string, int> capValues)
+        private void SendException(string operationName, string problemId, ExceptionTelemetry exceptionTelemetry, Exception exception)
         {
-            int temp;
-            if (capValues.TryGetValue(dimensionValue, out temp))
+            string exceptionKey;
+
+            if (string.IsNullOrEmpty(operationName) == true)
             {
-                return dimensionValue;
+                exceptionKey = problemId;
+            }
+            else
+            {
+                exceptionKey = operationName + problemId;
             }
-            else if (capValues.Count > 100)
+
+            if (this.exceptionKeyValues.Contains(exceptionKey) == true)
             {
-                return "OtherValue";
+                return;
+            }
+
+            this.exceptionKeyValues.RwLock.EnterWriteLock();
+
+            if (this.exceptionKeyValues.ValueCache.Contains(exceptionKey) == false)
+            {
+                this.exceptionKeyValues.ValueCache.Add(exceptionKey);
+
+                this.exceptionKeyValues.RwLock.ExitWriteLock();
+
+                if (exceptionTelemetry == null)
+                {
+                    exceptionTelemetry = new ExceptionTelemetry(exception);
+                    this.telemetryClient.Initialize(exceptionTelemetry);
+                }
+
+                StackTrace st = new StackTrace(3, true);
```

Comment:
Is there a way to ensure that no in-lining or other optimization happens and skipping 3 frames always leads to the consistent result? The previous methods do not seem to satisfy any of the in-lining/optimization requirements, so this should work as is unless those requirements change in CLR.